### ISSUE TITLE
Where should I put my database connect code ? 

### ISSUE THREAD

**Issue by ppzz** [ice-services/moleculer#175]
Hello guys .
I'm a starter for moleculer , I encounter some problem about database connection.

---
I try to connect database(MongoDB) in lifecycle event: `created`
and here is my code:
```js
async created () {
  const client = await MongoClient.connect(MONGO_URL)
  const db = client.db(MONGO_DB)
  this.groupCollection = db.collection(MONGO_GROUP_COLLECTION)
}
```
I want get a  usable collection after the service started.  but when I run my test file (see below). 
there is an Error occurred .it says  groupCollection is undefined (see below).
It seems like groupCollection isn't established when I use it .
```js
  const broker = new ServiceBroker()
  broker.createService(TestService)

  beforeAll(() => broker.start())
  afterAll(() => broker.stop())

  describe(`createGroup action`, () => {
    it(`should create a document in mongo db`, async () => {
      // cretateGroup will call this.groupCollection.insertOne()
      const objectID = await broker.call(`${serviceName}.createGroup`)
      expect(objectID.length).toBe(24)
    })
  })
```
```
   TypeError: Cannot read property 'insertOne' of undefined

      33 |
      34 |     createGroup (group) {
    > 35 |       return this.groupCollection.insertOne(group)
      36 |     },
      37 |
```

in fact, after  I checked source code of moleculer. I found that method `created` called directly, with no promise wrapped. but the method started is wrapped by promise. And I try to put the database initialization code into started event , then test is passed. 

1. Should I put database initialization code to started event？ 
2. Is there  a better way to get database connection? 
3. what  circumstances I need to use created event ?

**Comment by zhaoyao91** [comment]
Yeah, I thought this is a bug of the design of moleculer lifecycle events. There should be pairs of hooks of each stage, such as `beforeStart` and `afterStart`. Without such `beforeStart` hook, we cannot prepare the service before receiving actions. 

Worse, the `stopped` hook should be triggered after the service being detached, but by the code, it is actually triggered before that.

**Comment by icebob** [comment]
Hi @ppzz,
I don't recommend to connect DB in `created` because as you mentioned, it doesn't wait for a Promise. 
You should start connecting in the `started` event which is called at `broker.start`. But I recommend checking the code of `moleculer-db`: https://github.com/ice-services/moleculer-db/blob/master/packages/moleculer-db/src/index.js#L769

**Comment by zhaoyao91** [comment]
@icebob hi, thank you for reply. The problem is, if I connect db in `started`, the service is already started and ready to handle action by that time. If action comes in before connection established, won't it bomb?

**Comment by icebob** [comment]
It's true. I will check we can change the `created` to support Promise.

**Comment by tinchoz49** [comment]
Hi everyone! how are you? 

Maybe is a problem of name conventions. 

The `started` doesn't mean that the service is already started, it means that moleculer already calls the `broker.start()`. Each service of the related broker starts only when the promise returning by the `started` method is resolved.

You can check this by adding a delay on your `started` method service.

```javascript
{
...
	started() {
		return new Promise(resolve => setTimeout(resolve, 5000));
	}
}
```

So, I think if you do this:
```javascript
async started () {
  const client = await MongoClient.connect(MONGO_URL)
  const db = client.db(MONGO_DB)
  this.groupCollection = db.collection(MONGO_GROUP_COLLECTION)
}
```

The service is going to wait for the database connection before really start to work and receive remote actions.

**Comment by icebob** [comment]
@tinchoz49 Yes, it was true until v0.11. But now transit connects and after that the services start. [Src](https://github.com/ice-services/moleculer/blob/master/src/service-broker.js#L329). 
The reason is the [dependencies](https://github.com/ice-services/moleculer/issues/102). The dependencies blocking the started, but if we don't start transporter connecting earlier, it will be a deadlock.

**Comment by tinchoz49** [comment]
Ah, so you can execute actions before the `started` finish. I didn't know :smile: 

In that case, I agree with you in giving promise support in the `created` method. Make sense :)

**Comment by icebob** [comment]
Yes. Maybe I can do it without breaking changes.

**Comment by icebob** [comment]
It will be solved in the next release v0.12.
The `created` in untouched. But other nodes don't send requests while all local services are not started.
So you can use `started` to connect to DB and return with Promise or use async/await for connecting.
